src/Core/Context.php

```diff
@@ -0,0 +1,164 @@
+ *   // Do something within the context
+ * } finally {
+ *   // This makes sure that $prev will be current after this execution
+ *   Context::current()->detach($prev);
+ * }
+ * // Here, we are 100% sure $prev is the current context.
+ * ```
+ */
+class Context
+{
+    /**
+     * @var Context
+     */
+    private static $current;
+
+    /**
+     * @var array
+     */
+    private $values;
+
+    /**
+     * Creates a new Context.
+     *
+     * @param array $initialValues
+     */
+    public function __construct($initialValues = [])
+    {
+        $this->values = $initialValues;
+    }
+
+    /**
+     * Creates a new context with the given key/value set.
+     *
+     * @param string $key
+     * @param mixed $value
+     * @return Context
+     */
+    public function withValue($key, $value)
+    {
+        $copy = $this->values;
+        $copy[$key] = $value;
+        return new Context($copy);
+    }
+
+    /**
+     * Fetches the value for a given key in this context. Returns the provided
+     * default if not set.
+     *
+     * @param string $key
+     * @param mixed $default [optional]
+     * @return mixed
+     */
+    public function value($key, $default = null)
+    {
+        return array_key_exists($key, $this->values)
+            ? $this->values[$key]
+            : $default;
+    }
+
+    /**
+     * Attaches this context, thus entering a new scope within which this
+     * context is current(). The previously current context is returned.
+     *
+     * @return Context
+     */
+    public function attach()
+    {
+        $current = self::current();
+        self::setCurrent($this);
+        return $current;
+    }
+
+    /**
+     * Reverses an attach(), restoring the previous context and exiting the
+     * current scope.
+     *
+     * @param  Context $toAttach
+     */
+    public function detach(Context $toAttach)
+    {
+        if (self::current() !== $this) {
+            trigger_error('Unexpected context to detach.', E_USER_WARNING);
+        }
+
+        self::setCurrent($toAttach);
+    }
+
+    /**
+     * Returns all the contained data.
+     *
+     * @return array
+     */
+    public function values()
+    {
+        return $this->values;
+    }
+
+    /**
+     * Returns the context associated with the current scope, will never return
+     * null.
+     *
+     * @return Context
+     */
+    public static function current()
+    {
+        if (!self::$current) {
+            self::setCurrent(self::background());
+        }
+
+        return self::$current;
+    }
+
+    public static function setCurrent(Context $toAttach)
+    {
+        self::$current = $toAttach;
+    }
+
+    /**
+     * Returns an empty context.
+     *
+     * @return Context
+     */
+    public static function background()
+    {
+        return new Context();
+    }
+
+    /**
+     * Resets the context to an initial state. This is generally used only for
+     * testing.
+     */
+    public static function reset()
```

Comment:
Is it any way we can make this not public but still available for testing?

Reply:
Unfortunately, not really. I can add an `@intenral` annotation, but that's only for documentation (i.e. some doc generators may skip it).